// @angular-architects/native-federation/src/config.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.shareAngularLocales = exports.DEFAULT_SKIP_LIST = exports.shareAll = exports.share = exports.findRootTsConfigJson = exports.withNativeFederation = void 0;
var build_1 = require("@softarc/native-federation/build");
Object.defineProperty(exports, "withNativeFederation", { enumerable: true, get: function () { return build_1.withNativeFederation; } });
var build_2 = require("@softarc/native-federation/build");
Object.defineProperty(exports, "findRootTsConfigJson", { enumerable: true, get: function () { return build_2.findRootTsConfigJson; } });
Object.defineProperty(exports, "share", { enumerable: true, get: function () { return build_2.share; } });
Object.defineProperty(exports, "shareAll", { enumerable: true, get: function () { return build_2.shareAll; } });
Object.defineProperty(exports, "DEFAULT_SKIP_LIST", { enumerable: true, get: function () { return build_2.DEFAULT_SKIP_LIST; } });
var angular_locales_1 = require("./utils/angular-locales");
Object.defineProperty(exports, "shareAngularLocales", { enumerable: true, get: function () { return angular_locales_1.shareAngularLocales; } });
