// @angular-architects/native-federation/src/builders/build/builder.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.runBuilder = runBuilder;
const tslib_1 = require("tslib");
const fs = tslib_1.__importStar(require("fs"));
const mrmime = tslib_1.__importStar(require("mrmime"));
const path = tslib_1.__importStar(require("path"));
const build_1 = require("@angular/build");
const private_1 = require("@angular/build/private");
const architect_1 = require("@angular-devkit/architect");
const options_1 = require("@angular-devkit/build-angular/src/builders/dev-server/options");
const build_2 = require("@softarc/native-federation/build");
const angular_esbuild_adapter_1 = require("../../utils/angular-esbuild-adapter");
const fs_1 = require("fs");
const fstart_as_data_url_1 = require("../../tools/fstart-as-data-url");
const mem_resuts_1 = require("../../utils/mem-resuts");
const i18n_1 = require("../../utils/i18n");
const rebuild_events_1 = require("../../utils/rebuild-events");
const shared_mappings_plugin_1 = require("../../utils/shared-mappings-plugin");
const updateIndexHtml_1 = require("../../utils/updateIndexHtml");
const federation_build_notifier_1 = require("./federation-build-notifier");
const originalWrite = process.stderr.write.bind(process.stderr);
process.stderr.write = function (chunk, encodingOrCallback, callback) {
    const str = typeof chunk === 'string' ? chunk : chunk.toString();
    if (str.includes('vite:import-analysis') &&
        str.includes('es-module-shims.js')) {
        return true;
    }
    if (typeof encodingOrCallback === 'function') {
        return originalWrite(chunk, encodingOrCallback);
    }
    return originalWrite(chunk, encodingOrCallback, callback);
};
function _buildApplication(options, context, pluginsOrExtensions) {
    let extensions;
    if (pluginsOrExtensions && Array.isArray(pluginsOrExtensions)) {
        extensions = {
            codePlugins: pluginsOrExtensions,
        };
    }
    else {
        extensions = pluginsOrExtensions;
    }
    return (0, private_1.buildApplicationInternal)(options, context, extensions);
}
async function* runBuilder(nfOptions, context) {
    let target = (0, architect_1.targetFromTargetString)(nfOptions.target);
    let targetOptions = (await context.getTargetOptions(target));
    let builder = await context.getBuilderNameForTarget(target);
    if (builder === '@angular-devkit/build-angular:browser-esbuild') {
        build_2.logger.info('.: NATIVE FEDERATION - UPDATE NEEDED :.');
        build_2.logger.info('');
        build_2.logger.info("Since version 17.1, Native Federation uses Angular's");
        build_2.logger.info('Application-Builder and its Dev-Server.');
        build_2.logger.info('');
        build_2.logger.info('If you are sill on Angular 17.0.x, please update to');
        build_2.logger.info('Angular 17.1.x or downgrade to Native Federation 17.0.x.');
        build_2.logger.info('');
        build_2.logger.info('For working with Native Federation 17.1.x (recommented), ');
        build_2.logger.info('please update your project config, e.g. in angular.json');
        build_2.logger.info('');
        build_2.logger.info('This command performs the needed update for default configs:');
        build_2.logger.info('');
        build_2.logger.info('\tng g @angular-architects/native-federation:appbuilder');
        build_2.logger.info('');
        build_2.logger.info('You need to run it once per application to migrate');
        build_2.logger.info('Please find more information here: https://shorturl.at/gADJW');
        return;
    }
    /**
     * Explicitly defined as devServer or if the target contains "serve"
     */
    const runServer = typeof nfOptions.devServer !== 'undefined'
        ? !!nfOptions.devServer
        : target.target.includes('serve');
    let options = (await context.validateOptions(runServer
        ? {
            ...targetOptions,
            port: nfOptions.port || targetOptions['port'],
        }
        : targetOptions, builder));
    let serverOptions = null;
    const write = true;
    const watch = nfOptions.watch;
    if (options['buildTarget']) {
        serverOptions = await (0, options_1.normalizeOptions)(context, context.target.project, options);
        target = (0, architect_1.targetFromTargetString)(options['buildTarget']);
        targetOptions = (await context.getTargetOptions(target));
        builder = await context.getBuilderNameForTarget(target);
        options = (await context.validateOptions(targetOptions, builder));
    }
    options.watch = watch;
    if (nfOptions.baseHref) {
        options.baseHref = nfOptions.baseHref;
    }
    if (nfOptions.outputPath) {
        options.outputPath = nfOptions.outputPath;
    }
    const rebuildEvents = new rebuild_events_1.RebuildHubs();
    const adapter = (0, angular_esbuild_adapter_1.createAngularBuildAdapter)(options, context, rebuildEvents);
    (0, build_2.setBuildAdapter)(adapter);
    (0, build_2.setLogLevel)(options.verbose ? 'verbose' : 'info');
    if (!options.outputPath) {
        options.outputPath = `dist/${context.target.project}`;
    }
    const outputPath = options.outputPath;
    const outputOptions = {
        browser: 'browser',
        server: 'server',
        media: 'media',
        ...(typeof outputPath === 'string' ? undefined : outputPath),
        base: typeof outputPath === 'string' ? outputPath : outputPath.base,
    };
    const i18n = await (0, i18n_1.getI18nConfig)(context);
    const localeFilter = getLocaleFilter(options, runServer);
    const sourceLocaleSegment = typeof i18n?.sourceLocale === 'string'
        ? i18n.sourceLocale
        : i18n?.sourceLocale?.subPath || i18n?.sourceLocale?.code || '';
    const browserOutputPath = path.join(outputOptions.base, outputOptions.browser, options.localize ? sourceLocaleSegment : '');
    const differentDevServerOutputPath = Array.isArray(localeFilter) && localeFilter.length === 1;
    const devServerOutputPath = !differentDevServerOutputPath
        ? browserOutputPath
        : path.join(outputOptions.base, outputOptions.browser, options.localize[0]);
    const entryPoint = path.join(path.dirname(options.tsConfig), 'src/main.ts');
    const fedOptions = {
        workspaceRoot: context.workspaceRoot,
        outputPath: browserOutputPath,
        federationConfig: inferConfigPath(options.tsConfig),
        tsConfig: options.tsConfig,
        verbose: options.verbose,
        watch: false, // options.watch,
        dev: !!nfOptions.dev,
        entryPoint,
        buildNotifications: nfOptions.buildNotifications,
        cacheExternalArtifacts: nfOptions.cacheExternalArtifacts,
    };
    const activateSsr = nfOptions.ssr && !nfOptions.dev;
    const start = process.hrtime();
    const config = await (0, build_2.loadFederationConfig)(fedOptions);
    build_2.logger.measure(start, 'To load the federation config.');
    const externals = (0, build_2.getExternals)(config);
    const plugins = [
        (0, shared_mappings_plugin_1.createSharedMappingsPlugin)(config.sharedMappings),
        {
            name: 'externals',
            setup(build) {
                if (!activateSsr && build.initialOptions.platform !== 'node') {
                    build.initialOptions.external = externals.filter((e) => e !== 'tslib');
                }
            },
        },
    ];
    // SSR build fails when externals are provided via the plugin
    if (activateSsr) {
        options.externalDependencies = externals;
    }
    const isLocalDevelopment = runServer && nfOptions.dev;
    // Initialize SSE reloader only for local development
    if (isLocalDevelopment && nfOptions.buildNotifications?.enable) {
        federation_build_notifier_1.federationBuildNotifier.initialize(nfOptions.buildNotifications.endpoint);
    }
    const middleware = [
        ...(isLocalDevelopment
            ? [
                federation_build_notifier_1.federationBuildNotifier.createEventMiddleware((req) => removeBaseHref(req, options.baseHref)),
            ]
            : []),
        (req, res, next) => {
            const url = removeBaseHref(req, options.baseHref);
            const fileName = path.join(fedOptions.workspaceRoot, devServerOutputPath, url);
            const exists = fs.existsSync(fileName);
            if (url !== '/' && url !== '' && exists) {
                const lookup = mrmime.lookup;
                const mimeType = lookup(path.extname(fileName)) || 'text/javascript';
                const rawBody = fs.readFileSync(fileName, 'utf-8');
                // TODO: Evaluate need for debug infos
                // const body = addDebugInformation(url, rawBody);
                const body = rawBody;
                res.writeHead(200, {
                    'Content-Type': mimeType,
                    'Access-Control-Allow-Origin': '*',
                    'Access-Control-Allow-Methods': 'GET, POST, PUT, DELETE',
                    'Access-Control-Allow-Headers': 'Content-Type',
                });
                res.end(body);
            }
            else {
                next();
            }
        },
    ];
    const memResults = new mem_resuts_1.MemResults();
    let first = true;
    let lastResult;
    if ((0, fs_1.existsSync)(fedOptions.outputPath)) {
        (0, fs_1.rmSync)(fedOptions.outputPath, { recursive: true });
    }
    if (!(0, fs_1.existsSync)(fedOptions.outputPath)) {
        (0, fs_1.mkdirSync)(fedOptions.outputPath, { recursive: true });
    }
    if (!write) {
        (0, angular_esbuild_adapter_1.setMemResultHandler)((outFiles, outDir) => {
            const fullOutDir = outDir
                ? path.join(fedOptions.workspaceRoot, outDir)
                : null;
            memResults.add(outFiles.map((f) => new mem_resuts_1.EsBuildResult(f, fullOutDir)));
        });
    }
    let federationResult;
    try {
        const start = process.hrtime();
        federationResult = await (0, build_2.buildForFederation)(config, fedOptions, externals);
        build_2.logger.measure(start, 'To build the artifacts.');
    }
    catch (e) {
        build_2.logger.error(e?.message ?? 'Building the artifacts failed');
        process.exit(1);
    }
    if (activateSsr) {
        writeFstartScript(fedOptions);
    }
    const hasLocales = i18n?.locales && Object.keys(i18n.locales).length > 0;
    if (hasLocales && localeFilter) {
        const start = process.hrtime();
        (0, i18n_1.translateFederationArtefacts)(i18n, localeFilter, outputOptions.base, federationResult);
        build_2.logger.measure(start, 'To translate the artifacts.');
    }
    options.deleteOutputPath = false;
    const appBuilderName = '@angular/build:application';
    const builderRun = runServer
        ? (0, private_1.serveWithVite)(serverOptions, appBuilderName, _buildApplication, context, nfOptions.skipHtmlTransform
            ? {}
            : { indexHtml: transformIndexHtml(nfOptions) }, {
            buildPlugins: plugins,
            middleware,
        })
        : (0, build_1.buildApplication)(options, context, {
            codePlugins: plugins,
            indexHtmlTransformer: transformIndexHtml(nfOptions),
        });
    const rebuildQueue = new build_2.RebuildQueue();
    try {
        for await (const output of builderRun) {
            lastResult = output;
            if (!write && output['outputFiles']) {
                memResults.add(output['outputFiles'].map((file) => new mem_resuts_1.EsBuildResult(file)));
            }
            if (!write && output['assetFiles']) {
                memResults.add(output['assetFiles'].map((file) => new mem_resuts_1.NgCliAssetResult(file)));
            }
            // if (write && !runServer && !nfOptions.skipHtmlTransform) {
            //   updateIndexHtml(fedOptions, nfOptions);
            // }
            // if (!runServer) {
            //   yield output;
            // }
            if (!first && (nfOptions.dev || watch)) {
                rebuildQueue
                    .enqueue(async (signal) => {
                    if (signal?.aborted) {
                        throw new build_2.AbortedError('Build canceled before starting');
                    }
                    await new Promise((resolve, reject) => {
                        const timeout = setTimeout(resolve, Math.max(10, nfOptions.rebuildDelay));
                        if (signal) {
                            const abortHandler = () => {
                                clearTimeout(timeout);
                                reject(new build_2.AbortedError('[builder] During delay.'));
                            };
                            signal.addEventListener('abort', abortHandler, { once: true });
                        }
                    });
                    if (signal?.aborted) {
                        throw new build_2.AbortedError('[builder] Before federation build.');
                    }
                    const start = process.hrtime();
                    federationResult = await (0, build_2.buildForFederation)(config, fedOptions, externals, {
                        skipMappingsAndExposed: false,
                        skipShared: true,
                        signal,
                    });
                    if (signal?.aborted) {
                        throw new build_2.AbortedError('[builder] After federation build.');
                    }
                    if (hasLocales && localeFilter) {
                        (0, i18n_1.translateFederationArtefacts)(i18n, localeFilter, outputOptions.base, federationResult);
                    }
                    if (signal?.aborted) {
                        throw new build_2.AbortedError('[builder] After federation translations.');
                    }
                    build_2.logger.info('Done!');
                    if (isLocalDevelopment) {
                        federation_build_notifier_1.federationBuildNotifier.broadcastBuildCompletion();
                    }
                    build_2.logger.measure(start, 'To rebuild the federation artifacts.');
                })
                    .catch((error) => {
                    if (error instanceof build_2.AbortedError) {
                        build_2.logger.verbose('Rebuild was canceled. Cancellation point: ' + error?.message);
                        federation_build_notifier_1.federationBuildNotifier.broadcastBuildCancellation();
                    }
                    else {
                        build_2.logger.error('Federation rebuild failed!');
                        if (options.verbose)
                            console.error(error);
                        if (isLocalDevelopment) {
                            federation_build_notifier_1.federationBuildNotifier.broadcastBuildError(error);
                        }
                    }
                });
            }
            first = false;
        }
    }
    finally {
        rebuildQueue.dispose();
        if (isLocalDevelopment) {
            federation_build_notifier_1.federationBuildNotifier.stopEventServer();
        }
    }
    yield lastResult || { success: false };
}
function removeBaseHref(req, baseHref) {
    let url = req.url;
    if (baseHref && url.startsWith(baseHref)) {
        url = url.substr(baseHref.length);
    }
    return url;
}
function writeFstartScript(fedOptions) {
    const serverOutpath = path.join(fedOptions.outputPath, '../server');
    const fstartPath = path.join(serverOutpath, 'fstart.mjs');
    const buffer = Buffer.from(fstart_as_data_url_1.fstart, 'base64');
    fs.mkdirSync(serverOutpath, { recursive: true });
    fs.writeFileSync(fstartPath, buffer, 'utf-8');
}
function getLocaleFilter(options, runServer) {
    let localize = options.localize || false;
    if (runServer && Array.isArray(localize) && localize.length > 1) {
        localize = false;
    }
    if (runServer && localize === true) {
        localize = false;
    }
    return localize;
}
function inferConfigPath(tsConfig) {
    const relProjectPath = path.dirname(tsConfig);
    const relConfigPath = path.join(relProjectPath, 'federation.config.js');
    return relConfigPath;
}
function transformIndexHtml(nfOptions) {
    return (content) => Promise.resolve((0, updateIndexHtml_1.updateScriptTags)(content, 'main.js', 'polyfills.js', nfOptions));
}
function addDebugInformation(fileName, rawBody) {
    if (fileName !== '/remoteEntry.json') {
        return rawBody;
    }
    const remoteEntry = JSON.parse(rawBody);
    const shared = remoteEntry.shared;
    if (!shared) {
        return rawBody;
    }
    const sharedForVite = shared.map((s) => ({
        ...s,
        packageName: `/@id/${s.packageName}`,
    }));
    remoteEntry.shared = [...shared, ...sharedForVite];
    return JSON.stringify(remoteEntry, null, 2);
}
// eslint-disable-next-line @typescript-eslint/no-explicit-any
exports.default = (0, architect_1.createBuilder)(runBuilder);
