// @angular-architects/native-federation/src/generators/native-federation/generator.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = default_1;
const tslib_1 = require("tslib");
const devkit_1 = require("@nx/devkit");
const path = tslib_1.__importStar(require("path"));
function normalizeOptions(tree, options) {
    const name = (0, devkit_1.names)(options.name).fileName;
    const projectDirectory = options.directory
        ? `${(0, devkit_1.names)(options.directory).fileName}/${name}`
        : name;
    const projectName = projectDirectory.replace(new RegExp('/', 'g'), '-');
    const projectRoot = `${(0, devkit_1.getWorkspaceLayout)(tree).libsDir}/${projectDirectory}`;
    const parsedTags = options.tags
        ? options.tags.split(',').map((s) => s.trim())
        : [];
    return {
        ...options,
        projectName,
        projectRoot,
        projectDirectory,
        parsedTags,
    };
}
function addFiles(tree, options) {
    const templateOptions = {
        ...options,
        ...(0, devkit_1.names)(options.name),
        offsetFromRoot: (0, devkit_1.offsetFromRoot)(options.projectRoot),
        template: '',
    };
    (0, devkit_1.generateFiles)(tree, path.join(__dirname, 'files'), options.projectRoot, templateOptions);
}
async function default_1(tree, options) {
    const normalizedOptions = normalizeOptions(tree, options);
    (0, devkit_1.addProjectConfiguration)(tree, normalizedOptions.projectName, {
        root: normalizedOptions.projectRoot,
        projectType: 'library',
        sourceRoot: `${normalizedOptions.projectRoot}/src`,
        targets: {
            build: {
                executor: '@angular-architects/native-federation:build',
            },
        },
        tags: normalizedOptions.parsedTags,
    });
    addFiles(tree, normalizedOptions);
    await (0, devkit_1.formatFiles)(tree);
}
