// @softarc/native-federation/src/lib/utils/bundle-caching.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.cacheEntry = exports.getChecksum = exports.getFilename = exports.getCachePath = void 0;
const tslib_1 = require("tslib");
const path_1 = tslib_1.__importDefault(require("path"));
const fs_1 = tslib_1.__importDefault(require("fs"));
const crypto_1 = tslib_1.__importDefault(require("crypto"));
const logger_1 = require("../utils/logger");
const getCachePath = (workspaceRoot, project) => path_1.default.join(workspaceRoot, 'node_modules/.cache/native-federation', project);
exports.getCachePath = getCachePath;
const getFilename = (title, dev) => {
    const devSuffix = dev ? '-dev' : '';
    return `${title}${devSuffix}.meta.json`;
};
exports.getFilename = getFilename;
const getChecksum = (shared, dev) => {
    const denseExternals = Object.keys(shared)
        .sort()
        .reduce((clean, external) => {
        return (clean +
            ':' +
            external +
            (shared[external].version ? `@${shared[external].version}` : ''));
    }, 'deps');
    return crypto_1.default
        .createHash('sha256')
        .update(denseExternals + `:dev=${dev}`)
        .digest('hex');
};
exports.getChecksum = getChecksum;
const cacheEntry = (pathToCache, fileName) => ({
    getMetadata: (checksum) => {
        const metadataFile = path_1.default.join(pathToCache, fileName);
        if (!fs_1.default.existsSync(pathToCache) || !fs_1.default.existsSync(metadataFile))
            return undefined;
        const cachedResult = JSON.parse(fs_1.default.readFileSync(metadataFile, 'utf-8'));
        if (cachedResult.checksum !== checksum)
            return undefined;
        return cachedResult;
    },
    persist: (payload) => {
        fs_1.default.writeFileSync(path_1.default.join(pathToCache, fileName), JSON.stringify(payload), 'utf-8');
    },
    copyFiles: (fullOutputPath) => {
        const metadataFile = path_1.default.join(pathToCache, fileName);
        if (!fs_1.default.existsSync(metadataFile))
            throw new Error('Error copying artifacts to dist, metadata file could not be found.');
        const cachedResult = JSON.parse(fs_1.default.readFileSync(metadataFile, 'utf-8'));
        fs_1.default.mkdirSync(path_1.default.dirname(fullOutputPath), { recursive: true });
        cachedResult.files.forEach((file) => {
            const cachedFile = path_1.default.join(pathToCache, file);
            const distFileName = path_1.default.join(fullOutputPath, file);
            if (fs_1.default.existsSync(cachedFile)) {
                fs_1.default.copyFileSync(cachedFile, distFileName);
            }
        });
    },
    clear: () => {
        const metadataFile = path_1.default.join(pathToCache, fileName);
        if (!fs_1.default.existsSync(pathToCache)) {
            fs_1.default.mkdirSync(pathToCache, { recursive: true });
            logger_1.logger.debug(`Creating cache folder '${pathToCache}' for '${fileName}'.`);
            return;
        }
        if (!fs_1.default.existsSync(metadataFile))
            return;
        logger_1.logger.debug(`Purging cached bundle '${metadataFile}'.`);
        const cachedResult = JSON.parse(fs_1.default.readFileSync(metadataFile, 'utf-8'));
        cachedResult.files.forEach((file) => {
            const cachedFile = path_1.default.join(pathToCache, file);
            if (fs_1.default.existsSync(cachedFile))
                fs_1.default.unlinkSync(cachedFile);
        });
        fs_1.default.unlinkSync(metadataFile);
    },
});
exports.cacheEntry = cacheEntry;
