// @softarc/native-federation/src/lib/utils/logger.js
"use strict";
/* eslint-disable @typescript-eslint/no-explicit-any */
Object.defineProperty(exports, "__esModule", { value: true });
exports.setLogLevel = exports.logger = void 0;
const tslib_1 = require("tslib");
const chalk_1 = tslib_1.__importDefault(require("chalk"));
let verbose = false;
exports.logger = {
    warn: (msg) => console.warn(chalk_1.default.bgYellow.ansi256(15)(' WARN '), msg),
    error: (msg) => console.error(chalk_1.default.bgRed.ansi256(15)(' ERRR '), msg),
    notice: (msg) => console.log(chalk_1.default.bgYellowBright.black(' NOTE '), msg),
    info: (msg) => console.log(chalk_1.default.bgGreen.ansi256(15)(' INFO '), msg),
    verbose: (msg) => verbose && console.log(chalk_1.default.bgGreen.ansi256(15)(' DBG! '), msg),
    debug: (msg) => verbose && console.log(chalk_1.default.bgGreen.ansi256(15)(' DBG! '), msg),
    measure: (start, milestone) => {
        if (!verbose)
            return;
        const [totalSeconds, nanoseconds] = process.hrtime(start);
        const minutes = Math.floor(totalSeconds / 60);
        const seconds = totalSeconds % 60;
        const milliseconds = nanoseconds / 1000000;
        const msFormatted = milliseconds.toFixed(3);
        const timeStr = `${minutes.toString().padStart(2, '0')}:${seconds
            .toString()
            .padStart(2, '0')}:${msFormatted.padStart(7, '0')}ms`;
        console.log(chalk_1.default.bgGreen.ansi256(15)(' DBG! '), `${timeStr} - ${milestone}`);
    },
};
const setLogLevel = (level) => {
    verbose = level === 'verbose';
};
exports.setLogLevel = setLogLevel;
