// @softarc/native-federation/src/lib/utils/package-info.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isESMExport = void 0;
exports.findPackageJsonFiles = findPackageJsonFiles;
exports.expandFolders = expandFolders;
exports.getPackageInfo = getPackageInfo;
exports.getVersionMaps = getVersionMaps;
exports.getPackageJsonFiles = getPackageJsonFiles;
exports.findDepPackageJson = findDepPackageJson;
exports._getPackageInfo = _getPackageInfo;
const tslib_1 = require("tslib");
const fs = tslib_1.__importStar(require("fs"));
const path = tslib_1.__importStar(require("path"));
const logger_1 = require("./logger");
const normalize_1 = require("./normalize");
const isESMExport = (e) => {
    if (e === 'import' || e === 'module-sync')
        return true;
    // Common ESM conventions
    if (e === 'module' || e === 'esm' || /^es20\d{2}$/.test(e))
        return true;
    if (e === 'require')
        return false;
    // Common CJS conventions
    if (e === 'cjs' || e === 'commonjs')
        return false;
    // Ambiguous
    return undefined;
};
exports.isESMExport = isESMExport;
const packageCache = {};
function findPackageJsonFiles(project, workspace) {
    return expandFolders(project, workspace)
        .map((f) => path.join(f, 'package.json'))
        .filter((f) => fs.existsSync(f));
}
function expandFolders(child, parent) {
    const result = [];
    parent = (0, normalize_1.normalize)(parent, true);
    child = (0, normalize_1.normalize)(child, true);
    if (!child.startsWith(parent)) {
        throw new Error(`Workspace folder ${path} needs to be a parent of the project folder ${child}`);
    }
    let current = child;
    while (current !== parent) {
        result.push(current);
        const cand = (0, normalize_1.normalize)(path.dirname(current), true);
        if (cand === current) {
            break;
        }
        current = cand;
    }
    result.push(parent);
    return result;
}
function getPackageInfo(packageName, workspaceRoot) {
    workspaceRoot = (0, normalize_1.normalize)(workspaceRoot, true);
    const packageJsonInfos = getPackageJsonFiles(workspaceRoot, workspaceRoot);
    for (const info of packageJsonInfos) {
        const cand = _getPackageInfo(packageName, info.directory);
        if (cand) {
            return cand;
        }
    }
    logger_1.logger.warn('No meta data found for shared lib ' + packageName);
    return null;
}
function getVersionMapCacheKey(project, workspace) {
    return `${project}**${workspace}`;
}
function getVersionMaps(project, workspace) {
    return getPackageJsonFiles(project, workspace).map((json) => ({
        ...json.content['dependencies'],
    }));
}
function getPackageJsonFiles(project, workspace) {
    const cacheKey = getVersionMapCacheKey(project, workspace);
    let maps = packageCache[cacheKey];
    if (maps) {
        return maps;
    }
    maps = findPackageJsonFiles(project, workspace).map((f) => {
        const content = JSON.parse(fs.readFileSync(f, 'utf-8'));
        const directory = (0, normalize_1.normalize)(path.dirname(f), true);
        const result = {
            content,
            directory,
        };
        return result;
    });
    packageCache[cacheKey] = maps;
    return maps;
}
function findDepPackageJson(packageName, projectRoot) {
    const mainPkgName = getPkgFolder(packageName);
    let mainPkgPath = path.join(projectRoot, 'node_modules', mainPkgName);
    let mainPkgJsonPath = path.join(mainPkgPath, 'package.json');
    let directory = projectRoot;
    while (path.dirname(directory) !== directory) {
        if (fs.existsSync(mainPkgJsonPath)) {
            break;
        }
        directory = (0, normalize_1.normalize)(path.dirname(directory), true);
        mainPkgPath = path.join(directory, 'node_modules', mainPkgName);
        mainPkgJsonPath = path.join(mainPkgPath, 'package.json');
    }
    if (!fs.existsSync(mainPkgJsonPath)) {
        // TODO: Add logger
        // context.logger.warn('No package.json found for ' + packageName);
        logger_1.logger.verbose('No package.json found for ' + packageName + ' in ' + mainPkgPath);
        return null;
    }
    return mainPkgJsonPath;
}
function replaceGlob(target, replacement) {
    if (!target)
        return undefined;
    if (typeof target === 'string')
        return target.replace('*', replacement);
    return Object.entries(target).reduce((a, [k, v]) => ({
        ...a,
        [k]: replaceGlob(v, replacement),
    }), {});
}
function findOptimalExport(target, info, isESM = undefined) {
    if (typeof target === 'string') {
        return {
            ...info,
            entryPoint: path.join(info.entryPoint, target),
            esm: isESM ?? info.esm,
        };
    }
    if (!target)
        return undefined;
    if (Array.isArray(target))
        return findOptimalExport(target[0], info, isESM);
    const exportTypes = Object.keys(target);
    // We prefer ESM exports for native support.
    if (typeof isESM === 'undefined') {
        const esmExport = exportTypes.find((e) => (0, exports.isESMExport)(e));
        if (esmExport) {
            return findOptimalExport(target[esmExport], info, true);
        }
    }
    // Node.js looks at the exports object and uses the first key that matches the current environment.
    const secondBestEntry = 'default' in target && target['default']
        ? 'default'
        : exportTypes.filter((e) => e !== 'types')[0];
    const secondBestExport = target[secondBestEntry];
    return findOptimalExport(secondBestExport, info, isESM ?? (0, exports.isESMExport)(secondBestEntry));
}
function _getPackageInfo(packageName, directory) {
    const mainPkgName = getPkgFolder(packageName);
    const mainPkgJsonPath = findDepPackageJson(packageName, directory);
    if (!mainPkgJsonPath) {
        return null;
    }
    const mainPkgPath = path.dirname(mainPkgJsonPath);
    const mainPkgJson = readJson(mainPkgJsonPath);
    const version = mainPkgJson['version'];
    const esm = mainPkgJson['type'] === 'module';
    if (!version) {
        logger_1.logger.warn('No version found for ' + packageName);
        return null;
    }
    const pathToSecondary = path.relative(mainPkgName, packageName);
    const relSecondaryPath = !pathToSecondary
        ? '.'
        : './' + pathToSecondary.replace(/\\/g, '/');
    let secondaryEntryPoint = undefined;
    const packageJsonExportsEntry = Object.keys(mainPkgJson?.exports ?? []).find((e) => {
        if (e === relSecondaryPath)
            return true;
        if (e === './*')
            return true;
        if (!e.endsWith('*'))
            return false;
        const globPath = e.substring(0, e.length - 1);
        return relSecondaryPath.startsWith(globPath);
    });
    if (packageJsonExportsEntry) {
        secondaryEntryPoint = mainPkgJson?.exports?.[packageJsonExportsEntry];
        if (packageJsonExportsEntry.endsWith('*')) {
            const replacement = relSecondaryPath.substring(packageJsonExportsEntry.length - 1);
            secondaryEntryPoint = replaceGlob(secondaryEntryPoint, replacement);
        }
    }
    if (secondaryEntryPoint) {
        const info = findOptimalExport(secondaryEntryPoint, {
            entryPoint: mainPkgPath,
            packageName,
            version,
            esm,
        });
        if (info)
            return info;
    }
    if (mainPkgJson['module'] && relSecondaryPath === '.') {
        return {
            entryPoint: path.join(mainPkgPath, mainPkgJson['module']),
            packageName,
            version,
            esm: true,
        };
    }
    const secondaryPgkPath = path.join(mainPkgPath, relSecondaryPath);
    const secondaryPgkJsonPath = path.join(secondaryPgkPath, 'package.json');
    let secondaryPgkJson = null;
    if (fs.existsSync(secondaryPgkJsonPath)) {
        secondaryPgkJson = readJson(secondaryPgkJsonPath);
    }
    if (secondaryPgkJson && secondaryPgkJson.module) {
        return {
            entryPoint: path.join(secondaryPgkPath, secondaryPgkJson.module),
            packageName,
            version,
            esm: true,
        };
    }
    let cand = path.join(secondaryPgkPath, 'index.mjs');
    if (fs.existsSync(cand)) {
        return {
            entryPoint: cand,
            packageName,
            version,
            esm: true,
        };
    }
    if (secondaryPgkJson && secondaryPgkJson.main) {
        return {
            entryPoint: path.join(secondaryPgkPath, secondaryPgkJson.main),
            packageName,
            version,
            esm,
        };
    }
    cand = path.join(secondaryPgkPath, 'index.js');
    if (fs.existsSync(cand)) {
        return {
            entryPoint: cand,
            packageName,
            version,
            esm,
        };
    }
    cand = secondaryPgkPath + '.js';
    if (fs.existsSync(cand)) {
        return {
            entryPoint: cand,
            packageName,
            version,
            esm,
        };
    }
    cand = secondaryPgkPath + '.mjs';
    if (fs.existsSync(cand)) {
        return {
            entryPoint: cand,
            packageName,
            version,
            esm,
        };
    }
    logger_1.logger.warn('No entry point found for ' + packageName);
    logger_1.logger.warn("If you don't need this package, skip it in your federation.config.js or consider moving it into devDependencies in your package.json");
    return null;
}
function readJson(mainPkgJsonPath) {
    return JSON.parse(fs.readFileSync(mainPkgJsonPath, 'utf-8'));
}
function getPkgFolder(packageName) {
    const parts = packageName.split('/');
    let folder = parts[0];
    if (folder.startsWith('@')) {
        folder += '/' + parts[1];
    }
    return folder;
}
