// @softarc/native-federation/src/lib/utils/rebuild-queue.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RebuildQueue = void 0;
const logger_1 = require("./logger");
class RebuildQueue {
    activeBuilds = new Map();
    buildCounter = 0;
    async enqueue(rebuildFn) {
        const buildId = ++this.buildCounter;
        const pendingCancellations = Array.from(this.activeBuilds.values()).map((buildInfo) => {
            buildInfo.controller.abort();
            return buildInfo.buildFinished.promise;
        });
        if (pendingCancellations.length > 0) {
            logger_1.logger.info(`Aborting ${pendingCancellations.length} bundling task(s)..`);
        }
        if (pendingCancellations.length > 0) {
            await Promise.all(pendingCancellations);
        }
        let buildFinished;
        const completionPromise = new Promise((resolve) => {
            buildFinished = resolve;
        });
        const control = {
            controller: new AbortController(),
            buildFinished: {
                resolve: buildFinished,
                promise: completionPromise,
            },
        };
        this.activeBuilds.set(buildId, control);
        try {
            await rebuildFn(control.controller.signal);
        }
        finally {
            control.buildFinished.resolve();
            this.activeBuilds.delete(buildId);
        }
    }
    dispose() {
        for (const [_, buildInfo] of this.activeBuilds) {
            buildInfo.controller.abort();
        }
        this.activeBuilds.clear();
    }
}
exports.RebuildQueue = RebuildQueue;
