// @softarc/native-federation/src/lib/utils/resolve-wildcard-keys.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.resolveWildcardKeys = resolveWildcardKeys;
const tslib_1 = require("tslib");
const fast_glob_1 = tslib_1.__importDefault(require("fast-glob"));
function escapeRegex(str) {
    return str.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
}
// Convert package.json exports pattern to glob pattern
// * in exports means "one segment", but for glob we need **/* for deep matching
// Src: https://hirok.io/posts/package-json-exports#exposing-all-package-files
function convertExportsToGlob(pattern) {
    return pattern.replace(/(?<!\*)\*(?!\*)/g, '**/*');
}
function compilePattern(pattern) {
    const tokens = pattern.split(/(\*\*|\*)/);
    const regexParts = [];
    for (const token of tokens) {
        if (token === '*') {
            regexParts.push('(.*)');
        }
        else {
            regexParts.push(escapeRegex(token));
        }
    }
    return new RegExp(`^${regexParts.join('')}$`);
}
function withoutWildcard(template, wildcardValues) {
    const tokens = template.split(/(\*\*|\*)/);
    let result = '';
    let i = 0;
    for (const token of tokens) {
        if (token === '*') {
            result += wildcardValues[i++];
        }
        else {
            result += token;
        }
    }
    return result;
}
function resolveWildcardKeys(keyPattern, valuePattern, cwd) {
    const normalizedPattern = valuePattern.replace(/^\.?\/+/, '');
    const globPattern = convertExportsToGlob(normalizedPattern);
    const regex = compilePattern(normalizedPattern);
    const files = fast_glob_1.default.sync(globPattern, {
        cwd,
        onlyFiles: true,
        deep: Infinity,
    });
    const keys = [];
    for (const file of files) {
        const relPath = file.replace(/\\/g, '/').replace(/^\.\//, '');
        const wildcards = relPath.match(regex);
        if (!wildcards)
            continue;
        keys.push({
            key: withoutWildcard(keyPattern, wildcards.slice(1)),
            value: relPath,
        });
    }
    return keys;
}
