// @softarc/native-federation/src/lib/config/share-utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DEFAULT_SECONDARIES_SKIP_LIST = void 0;
exports.findRootTsConfigJson = findRootTsConfigJson;
exports.shareAll = shareAll;
exports.setInferVersion = setInferVersion;
exports.share = share;
const path = require("path");
const fs = require("fs");
const process_1 = require("process");
const default_skip_list_1 = require("../core/default-skip-list");
const package_info_1 = require("../utils/package-info");
const configuration_context_1 = require("./configuration-context");
const logger_1 = require("../utils/logger");
const resolve_wildcard_keys_1 = require("../utils/resolve-wildcard-keys");
let inferVersion = false;
exports.DEFAULT_SECONDARIES_SKIP_LIST = [
    '@angular/router/upgrade',
    '@angular/common/upgrade',
];
function findRootTsConfigJson() {
    const packageJson = findPackageJson((0, process_1.cwd)());
    const projectRoot = path.dirname(packageJson);
    const tsConfigBaseJson = path.join(projectRoot, 'tsconfig.base.json');
    const tsConfigJson = path.join(projectRoot, 'tsconfig.json');
    if (fs.existsSync(tsConfigBaseJson)) {
        return tsConfigBaseJson;
    }
    else if (fs.existsSync(tsConfigJson)) {
        return tsConfigJson;
    }
    throw new Error('Neither a tsconfig.json nor a tsconfig.base.json was found');
}
function findPackageJson(folder) {
    while (!fs.existsSync(path.join(folder, 'package.json')) &&
        path.dirname(folder) !== folder) {
        folder = path.dirname(folder);
    }
    const filePath = path.join(folder, 'package.json');
    if (fs.existsSync(filePath)) {
        return filePath;
    }
    throw new Error('no package.json found. Searched the following folder and all parents: ' +
        folder);
}
function lookupVersion(key, workspaceRoot) {
    const versionMaps = (0, package_info_1.getVersionMaps)(workspaceRoot, workspaceRoot);
    for (const versionMap of versionMaps) {
        const version = lookupVersionInMap(key, versionMap);
        if (version) {
            return version;
        }
    }
    throw new Error(`Shared Dependency ${key} has requiredVersion:'auto'. However, this dependency is not found in your package.json`);
}
function lookupVersionInMap(key, versions) {
    const parts = key.split('/');
    if (parts.length >= 2 && parts[0].startsWith('@')) {
        key = parts[0] + '/' + parts[1];
    }
    else {
        key = parts[0];
    }
    if (key.toLowerCase() === '@angular-architects/module-federation-runtime') {
        key = '@angular-architects/module-federation';
    }
    if (!versions[key]) {
        return null;
    }
    return versions[key];
}
function _findSecondaries(libPath, excludes, shareObject, acc, preparedSkipList) {
    const files = fs.readdirSync(libPath);
    const secondaries = files
        .map((f) => path.join(libPath, f))
        .filter((f) => fs.lstatSync(f).isDirectory() && !f.endsWith('node_modules'));
    for (const s of secondaries) {
        if (fs.existsSync(path.join(s, 'package.json'))) {
            const secondaryLibName = s
                .replace(/\\/g, '/')
                .replace(/^.*node_modules[/]/, '');
            const inCustomSkipList = excludes.some((e) => e === secondaryLibName ||
                (e.endsWith('*') && secondaryLibName.startsWith(e.slice(0, -1))));
            if (inCustomSkipList)
                continue;
            if ((0, default_skip_list_1.isInSkipList)(secondaryLibName, preparedSkipList)) {
                continue;
            }
            acc[secondaryLibName] = { ...shareObject };
        }
        _findSecondaries(s, excludes, shareObject, acc, preparedSkipList);
    }
}
function findSecondaries(libPath, excludes, shareObject, preparedSkipList) {
    const acc = {};
    _findSecondaries(libPath, excludes, shareObject, acc, preparedSkipList);
    return acc;
}
function getSecondaries(includeSecondaries, libPath, key, shareObject, preparedSkipList) {
    let exclude = [...exports.DEFAULT_SECONDARIES_SKIP_LIST];
    let resolveGlob = false;
    if (typeof includeSecondaries === 'object') {
        if (Array.isArray(includeSecondaries.skip)) {
            exclude = includeSecondaries.skip;
        }
        else if (typeof includeSecondaries.skip === 'string') {
            exclude = [includeSecondaries.skip];
        }
        resolveGlob = !!includeSecondaries.resolveGlob;
    }
    // const libPath = path.join(path.dirname(packagePath), 'node_modules', key);
    if (!fs.existsSync(libPath)) {
        return {};
    }
    const configured = readConfiguredSecondaries(key, libPath, exclude, shareObject, preparedSkipList, resolveGlob);
    if (configured) {
        return configured;
    }
    // Fallback: Search folders
    const secondaries = findSecondaries(libPath, exclude, shareObject, preparedSkipList);
    return secondaries;
}
function readConfiguredSecondaries(parent, libPath, exclude, shareObject, preparedSkipList, resolveGlob) {
    const libPackageJson = path.join(libPath, 'package.json');
    if (!fs.existsSync(libPackageJson)) {
        return null;
    }
    const packageJson = JSON.parse(fs.readFileSync(libPackageJson, 'utf-8'));
    const version = packageJson['version'];
    const esm = packageJson['type'] === 'module';
    const exports = packageJson['exports'];
    if (!exports) {
        return null;
    }
    const keys = Object.keys(exports).filter((key) => key != '.' &&
        key != './package.json' &&
        key.startsWith('./') &&
        (exports[key]['default'] ||
            exports[key]['import'] ||
            typeof exports[key] === 'string'));
    const result = {};
    const discoveredFiles = new Set();
    for (const key of keys) {
        const secondaryName = path.join(parent, key).replace(/\\/g, '/');
        const inCustomSkipList = exclude.some((e) => e === secondaryName ||
            (e.endsWith('*') && secondaryName.startsWith(e.slice(0, -1))));
        if (inCustomSkipList)
            continue;
        if ((0, default_skip_list_1.isInSkipList)(secondaryName, preparedSkipList)) {
            continue;
        }
        const entry = getDefaultEntry(exports, key);
        if (typeof entry !== 'string') {
            console.log('No entry point found for ' + secondaryName);
            continue;
        }
        if (!entry?.endsWith('.js') &&
            !entry?.endsWith('.mjs') &&
            !entry?.endsWith('.cjs')) {
            continue;
        }
        const items = resolveGlobSecondaries(key, libPath, parent, secondaryName, entry, { discovered: discoveredFiles, skip: exclude }, resolveGlob);
        items.forEach((e) => discoveredFiles.add(typeof e === 'string' ? e : e.value));
        for (const item of items) {
            if (typeof item === 'object') {
                result[item.key] = {
                    ...shareObject,
                    packageInfo: {
                        entryPoint: item.value,
                        version: shareObject.version ?? version,
                        esm,
                    },
                };
            }
            else {
                result[item] = {
                    ...shareObject,
                };
            }
        }
    }
    return result;
}
function resolveGlobSecondaries(key, libPath, parent, secondaryName, entry, excludes, resolveGlob) {
    let items = [];
    if (key.includes('*')) {
        if (!resolveGlob)
            return items;
        const expanded = (0, resolve_wildcard_keys_1.resolveWildcardKeys)(key, entry, libPath);
        items = expanded
            .map((e) => ({
            key: path.join(parent, e.key),
            value: path.join(libPath, e.value),
        }))
            .filter((i) => {
            if (excludes.skip.some((e) => e.endsWith('*') ? i.key.startsWith(e.slice(0, -1)) : e === i.key)) {
                return false;
            }
            if (excludes.discovered.has(i.value)) {
                return false;
            }
            return true;
        });
    }
    else {
        items = [secondaryName];
    }
    return items;
}
function getDefaultEntry(exports, key) {
    let entry = '';
    if (typeof exports[key] === 'string') {
        entry = exports[key];
    }
    if (!entry) {
        entry = exports[key]?.['default'];
        if (typeof entry === 'object') {
            entry = entry['default'];
        }
    }
    if (!entry) {
        entry = exports[key]?.['import'];
        if (typeof entry === 'object') {
            entry = entry['import'] ?? entry['default'];
        }
    }
    if (!entry) {
        entry = exports[key]?.['require'];
        if (typeof entry === 'object') {
            entry = entry['require'] ?? entry['default'];
        }
    }
    return entry;
}
function shareAll(config = {}, opts = {}) {
    // let workspacePath: string | undefined = undefined;
    const projectPath = inferProjectPath(opts.projectPath);
    // workspacePath = getConfigContext().workspaceRoot ?? '';
    // if (!workspacePath) {
    //   workspacePath = projectPath;
    // }
    const versionMaps = (0, package_info_1.getVersionMaps)(projectPath, projectPath);
    const sharedExternals = {};
    const preparedSkipList = (0, default_skip_list_1.prepareSkipList)(opts.skipList ?? default_skip_list_1.DEFAULT_SKIP_LIST);
    for (const versions of versionMaps) {
        for (const key in versions) {
            if ((0, default_skip_list_1.isInSkipList)(key, preparedSkipList)) {
                continue;
            }
            if (!!opts.overrides &&
                Object.keys(opts.overrides).some((o) => key.startsWith(o))) {
                continue;
            }
            const inferVersion = !config.requiredVersion || config.requiredVersion === 'auto';
            const requiredVersion = inferVersion
                ? versions[key]
                : config.requiredVersion;
            if (!sharedExternals[key]) {
                sharedExternals[key] = { ...config, requiredVersion };
            }
        }
    }
    return {
        ...share(sharedExternals, opts.projectPath, opts.skipList ?? default_skip_list_1.DEFAULT_SKIP_LIST),
        ...(!opts.overrides
            ? {}
            : share(opts.overrides, opts.projectPath, opts.skipList ?? default_skip_list_1.DEFAULT_SKIP_LIST)),
    };
}
function inferProjectPath(projectPath) {
    if (!projectPath && (0, configuration_context_1.getConfigContext)().packageJson) {
        projectPath = path.dirname((0, configuration_context_1.getConfigContext)().packageJson || '');
    }
    if (!projectPath && (0, configuration_context_1.getConfigContext)().workspaceRoot) {
        projectPath = (0, configuration_context_1.getConfigContext)().workspaceRoot || '';
    }
    if (!projectPath) {
        projectPath = (0, process_1.cwd)();
    }
    return projectPath;
}
function setInferVersion(infer) {
    inferVersion = infer;
}
function findTransientDeps(configuredShareObjects, projectRoot, preparedSkipList) {
    const discovered = new Set();
    const result = [];
    const packageNames = Object.keys(configuredShareObjects);
    for (const packageName of packageNames) {
        const shareConfig = configuredShareObjects[packageName];
        if (typeof shareConfig === 'object' && shareConfig.transient) {
            logger_1.logger.warn('PLEASE NOTE: The transient flag in your federation.config.js');
            logger_1.logger.warn('is deprecated. Please remove it. Meanwhile, Native Federation');
            logger_1.logger.warn('uses the underlying bundler for splitting transient');
            logger_1.logger.warn('dependencies into separate chunks, _when_ necessary.');
            const packagePath = path.join(projectRoot, 'node_modules', packageName, 'package.json');
            _findTransientDeps(packagePath, projectRoot, preparedSkipList, discovered, result);
        }
    }
    return result;
}
function _findTransientDeps(packagePath, projectRoot, preparedSkipList, discovered, result) {
    if (!fs.existsSync(packagePath)) {
        return;
    }
    const packageJson = JSON.parse(fs.readFileSync(packagePath, 'utf-8'));
    const deps = Object.keys(packageJson.dependencies ?? {});
    for (const dep of deps) {
        const depPackageJson = path.join(projectRoot, 'node_modules', dep, 'package.json');
        const depPath = path.dirname(depPackageJson);
        if (!discovered.has(depPackageJson) &&
            !(0, default_skip_list_1.isInSkipList)(dep, preparedSkipList) &&
            fs.existsSync(depPackageJson)) {
            discovered.add(depPackageJson);
            const version = packageJson.dependencies[dep];
            result.push({
                packageName: dep,
                requiredVersion: version,
                packagePath: depPath,
            });
            _findTransientDeps(depPackageJson, projectRoot, preparedSkipList, discovered, result);
        }
    }
}
function share(configuredShareObjects, projectPath = '', skipList = default_skip_list_1.DEFAULT_SKIP_LIST) {
    projectPath = inferProjectPath(projectPath);
    const packagePath = findPackageJson(projectPath);
    const packageDirectory = path.dirname(packagePath);
    const preparedSkipList = (0, default_skip_list_1.prepareSkipList)(skipList);
    const transientDeps = findTransientDeps(configuredShareObjects, packageDirectory, preparedSkipList);
    const transientShareObject = transientDeps.reduce((acc, curr) => ({
        ...acc,
        [curr.packageName]: { path: curr.packagePath },
    }), {});
    const shareObjects = {
        ...configuredShareObjects,
        ...transientShareObject,
    };
    const result = {};
    let includeSecondaries;
    for (const key in shareObjects) {
        includeSecondaries = false;
        const shareObject = shareObjects[key];
        if (shareObject.requiredVersion === 'auto' ||
            (inferVersion && typeof shareObject.requiredVersion === 'undefined')) {
            const version = lookupVersion(key, projectPath);
            shareObject.requiredVersion = version;
            shareObject.version = version.replace(/^\D*/, '');
        }
        if (typeof shareObject.includeSecondaries === 'undefined') {
            shareObject.includeSecondaries = true;
        }
        if (shareObject.includeSecondaries) {
            includeSecondaries = shareObject.includeSecondaries;
            delete shareObject.includeSecondaries;
            if (includeSecondaries?.keepAll)
                shareObject.includeSecondaries = true;
        }
        result[key] = shareObject;
        if (includeSecondaries) {
            const libPackageJson = (0, package_info_1.findDepPackageJson)(key, path.dirname(packagePath));
            if (!libPackageJson) {
                logger_1.logger.error('Could not find folder containing dep ' + key);
                continue;
            }
            const libPath = path.dirname(libPackageJson);
            const secondaries = getSecondaries(includeSecondaries, libPath, key, shareObject, preparedSkipList);
            if (secondaries) {
                addSecondaries(secondaries, result);
            }
        }
    }
    return result;
}
function addSecondaries(secondaries, result) {
    for (const key in secondaries) {
        result[key] = secondaries[key];
    }
}
