// @softarc/native-federation/src/lib/config/with-native-federation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.withNativeFederation = withNativeFederation;
const mapped_paths_1 = require("../utils/mapped-paths");
const share_utils_1 = require("./share-utils");
const default_skip_list_1 = require("../core/default-skip-list");
const logger_1 = require("../utils/logger");
const default_server_deps_list_1 = require("../core/default-server-deps-list");
function withNativeFederation(config) {
    const skip = (0, default_skip_list_1.prepareSkipList)(config.skip ?? []);
    const normalized = {
        name: config.name ?? '',
        exposes: config.exposes ?? {},
        shared: normalizeShared(config, skip),
        sharedMappings: normalizeSharedMappings(config, skip),
        skip,
        externals: config.externals ?? [],
        features: {
            mappingVersion: config.features?.mappingVersion ?? false,
            ignoreUnusedDeps: config.features?.ignoreUnusedDeps ?? false,
        },
    };
    // This is for being backwards compatible
    if (!normalized.features.ignoreUnusedDeps) {
        normalized.shared = filterShared(normalized.shared);
    }
    return normalized;
}
function filterShared(shared) {
    const keys = Object.keys(shared).filter((k) => !k.startsWith('@angular/common/locales'));
    const filtered = keys.reduce((acc, curr) => ({
        ...acc,
        [curr]: shared[curr],
    }), {});
    return filtered;
}
function normalizeShared(config, skip) {
    let result = {};
    const shared = config.shared;
    if (!shared) {
        result = (0, share_utils_1.shareAll)({
            singleton: true,
            strictVersion: true,
            requiredVersion: 'auto',
            platform: 'browser',
        });
    }
    else {
        result = Object.keys(shared).reduce((acc, cur) => ({
            ...acc,
            [cur.replace(/\\/g, '/')]: {
                requiredVersion: shared[cur].requiredVersion ?? 'auto',
                singleton: shared[cur].singleton ?? false,
                strictVersion: shared[cur].strictVersion ?? false,
                version: shared[cur].version,
                includeSecondaries: shared[cur].includeSecondaries,
                packageInfo: shared[cur].packageInfo,
                platform: shared[cur].platform ?? getDefaultPlatform(cur),
                build: shared[cur].build ?? 'default',
            },
        }), {});
        //result = share(result) as Record<string, NormalizedSharedConfig>;
    }
    result = Object.keys(result)
        .filter((key) => !(0, default_skip_list_1.isInSkipList)(key, skip))
        .reduce((acc, cur) => ({
        ...acc,
        [cur]: result[cur],
    }), {});
    return result;
}
function normalizeSharedMappings(config, skip) {
    const rootTsConfigPath = (0, share_utils_1.findRootTsConfigJson)();
    const paths = (0, mapped_paths_1.getMappedPaths)({
        rootTsConfigPath,
        sharedMappings: config.sharedMappings,
    });
    const result = paths.filter((p) => !(0, default_skip_list_1.isInSkipList)(p.key, skip) && !p.key.includes('*'));
    if (paths.find((p) => p.key.includes('*'))) {
        logger_1.logger.warn('Sharing mapped paths with wildcards (*) not supported');
    }
    return result;
}
function getDefaultPlatform(cur) {
    if (default_server_deps_list_1.DEFAULT_SERVER_DEPS_LIST.find((e) => cur.startsWith(e))) {
        return 'node';
    }
    else {
        return 'browser';
    }
}
