// @softarc/native-federation/src/lib/core/build-for-federation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.defaultBuildParams = void 0;
exports.buildForFederation = buildForFederation;
const bundle_exposed_and_mappings_1 = require("./bundle-exposed-and-mappings");
const bundle_shared_1 = require("./bundle-shared");
const write_federation_info_1 = require("./write-federation-info");
const write_import_map_1 = require("./write-import-map");
const logger_1 = require("../utils/logger");
const bundle_caching_1 = require("./../utils/bundle-caching");
const normalize_1 = require("../utils/normalize");
const errors_1 = require("../utils/errors");
const config_utils_1 = require("../utils/config-utils");
exports.defaultBuildParams = {
    skipMappingsAndExposed: false,
    skipShared: false,
};
const sharedPackageInfoCache = [];
async function buildForFederation(config, fedOptions, externals, buildParams = exports.defaultBuildParams) {
    const signal = buildParams.signal;
    let artefactInfo;
    if (!buildParams.skipMappingsAndExposed) {
        const start = process.hrtime();
        artefactInfo = await (0, bundle_exposed_and_mappings_1.bundleExposedAndMappings)(config, fedOptions, externals, signal);
        logger_1.logger.measure(start, '[build artifacts] - To bundle all mappings and exposed.');
        if (signal?.aborted)
            throw new errors_1.AbortedError('[buildForFederation] After exposed-and-mappings bundle');
    }
    const exposedInfo = !artefactInfo
        ? (0, bundle_exposed_and_mappings_1.describeExposed)(config, fedOptions)
        : artefactInfo.exposes;
    const cacheProjectFolder = (0, config_utils_1.resolveProjectName)(config);
    const pathToCache = (0, bundle_caching_1.getCachePath)(fedOptions.workspaceRoot, cacheProjectFolder);
    if (!buildParams.skipShared && sharedPackageInfoCache.length > 0) {
        logger_1.logger.info('Checksum matched, re-using cached externals.');
    }
    if (!buildParams.skipShared && sharedPackageInfoCache.length === 0) {
        const { sharedBrowser, sharedServer, separateBrowser, separateServer } = splitShared(config.shared);
        if (Object.keys(sharedBrowser).length > 0) {
            notifyBundling('browser-shared');
            const start = process.hrtime();
            const sharedPackageInfoBrowser = await (0, bundle_shared_1.bundleShared)(sharedBrowser, config, fedOptions, externals, 'browser', { pathToCache, bundleName: 'browser-shared' });
            logger_1.logger.measure(start, '[build artifacts] - To bundle all shared browser externals');
            sharedPackageInfoCache.push(...sharedPackageInfoBrowser);
            if (signal?.aborted)
                throw new errors_1.AbortedError('[buildForFederation] After shared-browser bundle');
        }
        if (Object.keys(sharedServer).length > 0) {
            notifyBundling('browser-shared');
            const start = process.hrtime();
            const sharedPackageInfoServer = await (0, bundle_shared_1.bundleShared)(sharedServer, config, fedOptions, externals, 'node', { pathToCache, bundleName: 'node-shared' });
            logger_1.logger.measure(start, '[build artifacts] - To bundle all shared node externals');
            sharedPackageInfoCache.push(...sharedPackageInfoServer);
            if (signal?.aborted)
                throw new errors_1.AbortedError('[buildForFederation] After shared-node bundle');
        }
        if (Object.keys(separateBrowser).length > 0) {
            notifyBundling('browser-shared');
            const start = process.hrtime();
            const separatePackageInfoBrowser = await bundleSeparatePackages(separateBrowser, externals, config, fedOptions, 'browser', pathToCache);
            logger_1.logger.measure(start, '[build artifacts] - To bundle all separate browser externals');
            sharedPackageInfoCache.push(...separatePackageInfoBrowser);
            if (signal?.aborted)
                throw new errors_1.AbortedError('[buildForFederation] After separate-browser bundle');
        }
        if (Object.keys(separateServer).length > 0) {
            notifyBundling('browser-shared');
            const start = process.hrtime();
            const separatePackageInfoServer = await bundleSeparatePackages(separateServer, externals, config, fedOptions, 'node', pathToCache);
            logger_1.logger.measure(start, '[build artifacts] - To bundle all separate node externals');
            sharedPackageInfoCache.push(...separatePackageInfoServer);
        }
        if (signal?.aborted)
            throw new errors_1.AbortedError('[buildForFederation] After separate-node bundle');
    }
    const sharedMappingInfo = !artefactInfo
        ? (0, bundle_exposed_and_mappings_1.describeSharedMappings)(config, fedOptions)
        : artefactInfo.mappings;
    const sharedInfo = [...sharedPackageInfoCache, ...sharedMappingInfo];
    const buildNotificationsEndpoint = fedOptions.buildNotifications?.enable && fedOptions.dev
        ? fedOptions.buildNotifications?.endpoint
        : undefined;
    const federationInfo = {
        name: config.name,
        shared: sharedInfo,
        exposes: exposedInfo,
        buildNotificationsEndpoint,
    };
    (0, write_federation_info_1.writeFederationInfo)(federationInfo, fedOptions);
    (0, write_import_map_1.writeImportMap)(sharedInfo, fedOptions);
    return federationInfo;
}
function inferPackageFromSecondary(secondary) {
    const parts = secondary.split('/');
    if (secondary.startsWith('@') && parts.length >= 2) {
        return parts[0] + '/' + parts[1];
    }
    return parts[0];
}
async function bundleSeparatePackages(separateBrowser, externals, config, fedOptions, platform, pathToCache) {
    const groupedByPackage = {};
    for (const [key, shared] of Object.entries(separateBrowser)) {
        const packageName = shared.build === 'separate' ? key : inferPackageFromSecondary(key);
        if (!groupedByPackage[packageName]) {
            groupedByPackage[packageName] = {};
        }
        groupedByPackage[packageName][key] = shared;
    }
    const bundlePromises = Object.entries(groupedByPackage).map(async ([packageName, sharedGroup]) => {
        return (0, bundle_shared_1.bundleShared)(sharedGroup, config, fedOptions, externals.filter((e) => !e.startsWith(packageName)), platform, {
            pathToCache,
            bundleName: `${platform}-${(0, normalize_1.normalizePackageName)(packageName)}`,
        });
    });
    const buildResults = await Promise.all(bundlePromises);
    return buildResults.flat();
}
function notifyBundling(platform) {
    logger_1.logger.info('Preparing shared npm packages for the platform ' + platform);
    logger_1.logger.notice('This only needs to be done once, as results are cached');
    logger_1.logger.notice("Skip packages you don't want to share in your federation config");
}
function splitShared(shared) {
    const sharedServer = {};
    const sharedBrowser = {};
    const separateBrowser = {};
    const separateServer = {};
    for (const key in shared) {
        const obj = shared[key];
        if (obj.platform === 'node') {
            if (obj.build === 'default') {
                sharedServer[key] = obj;
            }
            else {
                separateServer[key] = obj;
            }
        }
        else if (obj.platform === 'browser') {
            if (obj.build === 'default') {
                sharedBrowser[key] = obj;
            }
            else {
                separateBrowser[key] = obj;
            }
        }
    }
    return {
        sharedBrowser,
        sharedServer,
        separateBrowser,
        separateServer,
    };
}
