// @softarc/native-federation/src/lib/core/federation-builder.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.federationBuilder = void 0;
const configuration_context_1 = require("../config/configuration-context");
const build_adapter_1 = require("./build-adapter");
const build_for_federation_1 = require("./build-for-federation");
const get_externals_1 = require("./get-externals");
const load_federation_config_1 = require("./load-federation-config");
let externals = [];
let config;
let fedOptions;
let fedInfo;
async function init(params) {
    (0, build_adapter_1.setBuildAdapter)(params.adapter);
    fedOptions = params.options;
    (0, configuration_context_1.useWorkspace)(params.options.workspaceRoot);
    (0, configuration_context_1.usePackageJson)(params.options.packageJson);
    config = await (0, load_federation_config_1.loadFederationConfig)(fedOptions);
    params.options.workspaceRoot =
        (0, configuration_context_1.getConfigContext)().workspaceRoot ?? params.options.workspaceRoot;
    externals = (0, get_externals_1.getExternals)(config);
}
async function build(buildParams = build_for_federation_1.defaultBuildParams) {
    fedInfo = await (0, build_for_federation_1.buildForFederation)(config, fedOptions, externals, buildParams);
}
exports.federationBuilder = {
    init,
    build,
    get federationInfo() {
        return fedInfo;
    },
    get externals() {
        return externals;
    },
    get config() {
        return config;
    },
};
