// @softarc/native-federation/src/lib/core/remove-unused-deps.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.removeUnusedDeps = removeUnusedDeps;
const tslib_1 = require("tslib");
const sheriff_core_1 = require("@softarc/sheriff-core");
const path_1 = tslib_1.__importDefault(require("path"));
const fs_1 = tslib_1.__importDefault(require("fs"));
const process_1 = require("process");
const package_info_1 = require("../utils/package-info");
const get_external_imports_1 = require("../utils/get-external-imports");
const normalize_1 = require("../utils/normalize");
const config_utils_1 = require("../utils/config-utils");
function removeUnusedDeps(config, main, workspaceRoot) {
    const fileInfos = (0, sheriff_core_1.getProjectData)(main, (0, process_1.cwd)(), {
        includeExternalLibraries: true,
    });
    const usedDeps = findUsedDeps(fileInfos, workspaceRoot, config);
    const usedPackageNames = usedDeps.usedPackageNames;
    const usedMappings = usedDeps.usedMappings;
    const projectName = (0, config_utils_1.resolveProjectName)(config);
    const usedPackageNamesWithTransient = addTransientDeps(usedPackageNames, workspaceRoot, projectName);
    const filteredShared = filterShared(config, usedPackageNamesWithTransient);
    return {
        ...config,
        shared: filteredShared,
        sharedMappings: [...usedMappings],
    };
}
function filterShared(config, usedPackageNamesWithTransient) {
    return Object.entries(config.shared)
        .filter(([shared, meta]) => !!meta.includeSecondaries || usedPackageNamesWithTransient.has(shared))
        .reduce((acc, [shared, meta]) => ({ ...acc, [shared]: meta }), {});
}
function findUsedDeps(fileInfos, workspaceRoot, config) {
    const usedPackageNames = new Set();
    const usedMappings = new Set();
    for (const fileName of Object.keys(fileInfos)) {
        const fileInfo = fileInfos[fileName];
        if (!fileInfo) {
            continue;
        }
        const libs = [
            ...(fileInfo.externalLibraries || []),
            ...(fileInfo.unresolvedImports || []),
        ];
        for (const pckg of libs) {
            usedPackageNames.add(pckg);
        }
        const fullFileName = path_1.default.join(workspaceRoot, fileName);
        const mappings = config.sharedMappings.filter((sm) => fullFileName.startsWith(sm.path));
        for (const mapping of mappings) {
            usedMappings.add(mapping);
        }
    }
    return { usedPackageNames, usedMappings };
}
function addTransientDeps(packages, workspaceRoot, projectName) {
    const packagesAndPeers = new Set([...packages]);
    const discovered = new Set(packagesAndPeers);
    const stack = [...packagesAndPeers];
    while (stack.length > 0) {
        const dep = stack.pop();
        if (!dep) {
            continue;
        }
        const pInfo = (0, package_info_1.getPackageInfo)(dep, workspaceRoot);
        if (!pInfo) {
            continue;
        }
        const peerDeps = getExternalImports(pInfo, workspaceRoot, projectName);
        for (const peerDep of peerDeps) {
            if (!discovered.has(peerDep)) {
                discovered.add(peerDep);
                stack.push(peerDep);
                packagesAndPeers.add(peerDep);
            }
        }
    }
    return packagesAndPeers;
}
function getExternalImports(pInfo, workspaceRoot, projectName) {
    const encodedPackageName = (0, normalize_1.normalizePackageName)(pInfo.packageName);
    const cacheFileName = `${encodedPackageName}-${pInfo.version}.deps.json`;
    const cachePath = path_1.default.join(workspaceRoot, 'node_modules/.cache/native-federation/_externals-metadata', projectName);
    const cacheFilePath = path_1.default.join(cachePath, cacheFileName);
    const cacheHit = fs_1.default.existsSync(cacheFilePath);
    let peerDeps;
    if (cacheHit) {
        peerDeps = JSON.parse(fs_1.default.readFileSync(cacheFilePath, 'utf-8'));
    }
    else {
        peerDeps = (0, get_external_imports_1.getExternalImports)(pInfo.entryPoint);
        fs_1.default.mkdirSync(cachePath, { recursive: true });
        fs_1.default.writeFileSync(cacheFilePath, JSON.stringify(peerDeps, undefined, 2), 'utf-8');
    }
    return peerDeps;
}
