// @angular-architects/native-federation/src/utils/mem-resuts.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MemResults = exports.NgCliAssetResult = exports.EsBuildResult = void 0;
const tslib_1 = require("tslib");
const fs = tslib_1.__importStar(require("fs"));
const path = tslib_1.__importStar(require("path"));
class EsBuildResult {
    outputFile;
    fullOutDir;
    get fileName() {
        if (this.fullOutDir) {
            return unify(path.relative(this.fullOutDir, this.outputFile.path));
        }
        else {
            return unify(this.outputFile.path);
        }
    }
    constructor(outputFile, fullOutDir) {
        this.outputFile = outputFile;
        this.fullOutDir = fullOutDir;
    }
    get() {
        return this.outputFile.contents;
    }
}
exports.EsBuildResult = EsBuildResult;
class NgCliAssetResult {
    assetFile;
    get fileName() {
        return unify(this.file.destination);
    }
    file;
    constructor(assetFile) {
        this.assetFile = assetFile;
        this.file = assetFile;
    }
    get() {
        return fs.readFileSync(this.file.source);
    }
}
exports.NgCliAssetResult = NgCliAssetResult;
class MemResults {
    map = new Map();
    add(result) {
        for (const file of result) {
            this.map.set(file.fileName, file);
        }
    }
    get(fileName) {
        return this.map.get(fileName);
    }
    getFileNames() {
        return [...this.map.keys()];
    }
}
exports.MemResults = MemResults;
function unify(path) {
    return path?.replace(/\\/g, '/');
}
