// @angular-architects/native-federation/src/utils/patch-angular-build.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.privateEntrySrc = void 0;
exports.patchAngularBuildPackageJson = patchAngularBuildPackageJson;
exports.patchAngularBuild = patchAngularBuild;
const tslib_1 = require("tslib");
const path = tslib_1.__importStar(require("path"));
const fs = tslib_1.__importStar(require("fs"));
exports.privateEntrySrc = `
exports = require('./src/private.js');
`;
function patchAngularBuildPackageJson(packageJson) {
    const exportsMap = packageJson['exports'];
    if (!exportsMap) {
        console.log('No need to patch @angular/build/package.json');
        return;
    }
    packageJson['_exports'] = exportsMap;
    delete packageJson['exports'];
    packageJson['types'] = './src/index.d.ts';
    packageJson['main'] = './src/index.js';
    packageJson['module'] = './src/index.js';
}
function patchAngularBuild(workspaceRoot) {
    const packagePath = path.join(workspaceRoot, 'node_modules/@angular/build/package.json');
    const privatePath = path.join(workspaceRoot, 'node_modules/@angular/build/private.js');
    if (!fs.existsSync(packagePath)) {
        return;
    }
    const packageJson = JSON.parse(fs.readFileSync(packagePath, 'utf-8'));
    patchAngularBuildPackageJson(packageJson);
    fs.writeFileSync(packagePath, JSON.stringify(packageJson, null, 2));
    fs.writeFileSync(privatePath, exports.privateEntrySrc);
    console.log('@angular/build/package.json patched');
}
